### DISCUSSION TITLE
Can I use browsh with a Firefox Fork?

### DISCUSSION THREAD

**Discussion by martin-braun** [browsh-org/browsh#506] [Q&A]
I want to test Floorp or Mullvad Browser with `browsh`. Is it possible, if so, how?

**Comment by tombh** [comment]
See the `--firefox.path string ` arg: https://www.brow.sh/docs/config/

**Reply by martin-braun** [reply]
Thanks for the quick response, I forgot to mention, does it work with Flatpak as well?

**Reply by tombh** [reply]
I honestly don't know, I've never used a Flatpak.

**Reply by martin-braun** [reply]
@tombh My first impression is `browsh` is still very experimental. I'm testing with Debian bookworm LXDE as host with the pre-bundled Firefox that comes with it.

It won't start via `--firefox.with-gui`, it's exceptionally slow in attaching an existing instance via `--firefox.use-existing`. But, without both modes and without Firefox running, it seems to manage to launch Firefox and attach itself pretty quick. It succeeds to start Firefox and from then on, things do work eventually.

In regards of Flatpak, I tinkled a bit with it:

```
path = ".local/share/flatpak/exports/bin/one.ablaze.floorp"
```

seemed to be the trick to get into the entry of the Floorp flatpak, a nice Firefox fork that I installed via `flatpak install one.ablaze.floorp --user`. `exports/bin/...` is linked by `flatpak` to the primary binary of the app. Things don't work though. They use their own version schema, thus `browsh` returns:

> Installed Firefox version 11.8.0 is too old. Firefox 57 or newer is needed.

Can I tell `browsh` to skip the version check?

**Reply by tombh** [reply]
I'm sorry, there's currently no way to skip the version check.

**Reply by martin-braun** [reply]
Alright, please lets address this enhancement request sometime: https://github.com/browsh-org/browsh/issues/508 (you can assign me to this issue, but I will have to look into that way later, unless of course you are interested in it to provide a flag sometime soon).

Have a nice evening.

**Reply by ypurpl** [reply]
Can I do this on Windows as well? doesnt seem to work for me. Getting `Could not find Firefox on your registry`

**Reply by martin-braun** [reply]
@tombh What about the idea to have a flag to bypass the Firefox validation checks altogether (version, registry, etc.) and try it anyways? cc @vvirtues

**Reply by tombh** [reply]
Sure, it makes sense. PRs are very welcome. I can't be sure when I'd get around to it myself.

**Reply by martin-braun** [reply]
@tombh I figured, there is no need to bypass the registry check on Windows. If someone wants to use a fork, obviously it is necessary to specify `firefox.path` so `browsh` can find the binary. Only when the argument's default value `firefox` is assigned to `firefox.path` it will attempt to resolve the path by doing `which firefox` on unix based systems or scan the registry for the path on Windows systems.

@vvirtues Your issue is not an issue, provide the path to the `exe` in the configuration. It most likely still won't work, unless the fork you are trying to use is passing the version check of `browsh`.

Stacktrace:
- https://github.com/browsh-org/browsh/blob/cb3ddd533fcdaa5c4caf87154a5b627b26d38ae7/interfacer/src/browsh/firefox_windows.go#L94
- https://github.com/browsh-org/browsh/blob/cb3ddd533fcdaa5c4caf87154a5b627b26d38ae7/interfacer/src/browsh/firefox_windows.go#L16
- https://github.com/browsh-org/browsh/blob/cb3ddd533fcdaa5c4caf87154a5b627b26d38ae7/interfacer/src/browsh/firefox.go#L111

**Reply by Compiux** [reply]
Windows user here, hoping this will provide some useful information
I'm trying to use Mullvad Browser. I updated the path in the configuration file (config.toml) removed the firefox profile to let Browsh create a new one (which it did successfully) and installed the extension in the Mullvad Browser (did this when I ran Browsh with the --firefox.with-gui option).
For the time being, the browser is stuck on the "Waiting for firefox to connect..." step.
I'm positive the path to the browser is correct (if I use the firefox.with-gui, the mullvad browser shows, but the CMD window is stuck)

This is my command:
`browsh_1.8.0_windows_amd64.exe --debug --firefox.path "C:\MullvadBrowser\Browser\mullvadbrowser.exe" --firefox.with-gui`

This is the debug file content:
`Looking in C:\Users\********\AppData\Roaming\browsh\config.toml for config.
Attempting to connect to Firefox Marionette
Starting Firefox in headless mode
Using Firefox at: C:\MullvadBrowser\Browser\mullvadbrowser.exe
Windows registry Firefox version: 130.0
Using default profile at: C:\Users\hicoreas\AppData\Roaming\browsh\firefox_profile
FF-CONSOLE: Crash Annotation GraphicsCriticalError: |[0][GFX1-]: RenderCompositorSWGL failed mapping default framebuffer, no dt (t=3.69696) [GFX1-]: RenderCompositorSWGL failed mapping default framebuffer, no dt
Sending `WebDriver:NewSession` to Firefox Marionette
Sending `Addon:Install` to Firefox Marionette
Starting Browsh CLI client
Starting websocket server...
FF-MRNT: 50:{"applicationType":"gecko","marionetteProtocol":3}
FF-MRNT: 1106:[1,1,{"error":"unknown error","message":"Expected absolute path: [Exception... \"Component returned failure code: 0x80520001 (NS_ERROR_FILE_UNRECOGNIZED_PATH) [nsIFile.initWithPath]\"  nsresult: \"0x80520001 (NS_ERROR_FILE_UNRECOGNIZED_PATH)\"  location: \"JS frame :: chrome://remote/content/marionette/addon.sys.mjs :: install :: line 63\"  data: no]","stacktrace":"RemoteError@chrome://remote/content/shared/RemoteError.sys.mjs:8:8\nWebDriverError@chrome://remote/content/shared/webdriver/Errors.sys.mjs:187:5\nUnknownError@chrome://remote/content/shared/webdriver/Errors.sys.mjs:736:5\ninstall@chrome://remote/content/marionette/addon.sys.mjs:65:13\nGeckoDriver.prototype.installAddon@chrome://remote/content/marionette/driver.sys.mjs:3003:21\ndespatch@chrome://remote/content/marionette/server.sys.mjs:320:40\nexecute@chrome://remote/content/marionette/server.sys.mjs:291:16\nonPacket/<@chrome://remote/content/marionette/server.sys.mjs:264:20\nonPacket@chrome://remote/content/marionette/server.sys.mjs:265:9\n_onJSONObjectReady/<@chrome://remote/content/marionette/transport.sys.mjs:496:20\n"},null]
Webextension not connected. Message not sent: /tty_size,117,30
Webextension not connected. Message not sent: /tty_size,117,30
Webextension not connected. Message not sent: /tty_size,117,30
FF-MRNT: 726:[1,0,null,{"sessionId":"f06f5b69-8f39-42c6-af6a-3abf017dac6b","capabilities":{"browserName":"mullvadbrowser","browserVersion":"115.14.0","platformName":"windows","acceptInsecureCerts":false,"pageLoadStrategy":"normal","setWindowRect":true,"timeouts":{"implicit":0,"pageLoad":300000,"script":30000},"strictFileInteractability":false,"unhandledPromptBehavior":"dismiss and notify","moz:accessibilityChecks":false,"moz:buildID":"20240805090000","moz:headless":true,"moz:platformVersion":"10.0","moz:processID":5040,"moz:profile":"C:\\Users\\hicoreas\\AppData\\Roaming\\browsh\\firefox_profile","moz:shutdownTimeout":60000,"moz:useNonSpecCompliantPointerOrigin":false,"moz:webdriverClick":true,"moz:windowless":false,"proxy":{}}}]
`
